hi-ml/src/health_ml/utils/split_dataset.py

```diff
@@ -0,0 +1,378 @@
+
+    def __getitem__(self, mode: ModelExecutionMode) -> pd.DataFrame:
+        if mode == ModelExecutionMode.TRAIN:
+            return self.train
+        elif mode == ModelExecutionMode.TEST:
+            return self.test
+        elif mode == ModelExecutionMode.VAL:
+            return self.val
+        else:
+            raise ValueError(f"Model execution mode not recognized: {mode}")
+
+    def restrict_subjects(self, restriction_pattern: str) -> DatasetSplits:
+        """
+        Creates a new dataset split that has at most the specified numbers of subjects in train, validation and test
+        sets respectively.
+
+        If `group_column` was specified, this operation may violate the grouping constraints and the resulting splits
+        object will have `group_column == None`.
+
+        :param restriction_pattern: a string containing zero or two commas, and otherwise digits or "+". An empty
+            substring will result in no restriction for the corresponding dataset. Thus "20,,3" means "restrict to 20
+            training images and 3 test images, with no restriction on validation". A "+" value means "reassign all
+            images from the set(s) with a numeric count (there must be at least one) to this set". Thus ",0,+" means
+            "leave the training set alone, but move all validation images to the test set", and "0,2,+" means "move
+            all training images and all but 2 validation images to the test set".
+        :return: A new dataset split object with (at most) the numbers of subjects specified by restrict_pattern
+        """
+
+        n_train, n_val, n_test = self.parse_restriction_pattern(restriction_pattern)
+
+        def restrict(df: pd.DataFrame, count: Optional[int]) -> Tuple[pd.DataFrame, pd.DataFrame]:
+            if count is None:  # Not specified: keep everything
+                return df, df[:0]
+            ids = df[self.subject_column].unique()
+            if count >= len(ids):  # "+", or a large number specified
+                return df, df[:0]
+            keep = ids[:count]
+            drop = ids[count:]
+            return df[df[self.subject_column].isin(keep)], df[df[self.subject_column].isin(drop)]
+
+        train, train_drop = restrict(self.train, n_train)
+        test, test_drop = restrict(self.test, n_test)
+        val, val_drop = restrict(self.val, n_val)
+        if n_train == sys.maxsize:
+            train = train.append(val_drop).append(test_drop)
+        elif n_test == sys.maxsize:
+            test = test.append(train_drop).append(val_drop)
+        elif n_val == sys.maxsize:
+            val = val.append(train_drop).append(test_drop)
+
+        return DatasetSplits(train=train, test=test, val=val, subject_column=self.subject_column, allow_empty=True)
+
+    @staticmethod
+    def parse_restriction_pattern(restriction_pattern: str) -> Tuple[Optional[int], Optional[int], Optional[int]]:
+        fields = restriction_pattern.split(",")
+
```

Comment:
Not sure we want all this logic about restricting subjects in hi-ml? Its main use was to allow for quicker training on smaller sets of data, but we are mostly using PL flags for that now.